File: cses/sliding_window/Sliding Window Cost.py
from sortedcontainers import SortedList
n,k=[int(x) for x in input().split()]
arr=[int(x) for x in input().split()]
class map:
    def __init__(self):
        self.a=SortedList()
        self.b=SortedList()
        self.a_sum=0
        self.b_sum=0
    def insert(self,val):
        if self.a and val>self.a[-1]:
            self.b.add(val)
            self.b_sum+=val
        else:
            self.a.add(val)
            self.a_sum+=val
        self.adjust()
    def remove(self,val):
        if val>self.a[-1]:
            self.b.remove(val)
            self.b_sum-=val
        else:
            self.a.remove(val)
            self.a_sum-=val
        self.adjust()

    def adjust(self):
        while len(self.b) > len(self.a):
            temp = self.b.pop(0)
            self.b_sum-=temp
            self.a.add(temp)
            self.a_sum+=temp
        while len(self.b) + 1 < len(self.a):
            temp = self.a.pop(-1)
            self.a_sum-=temp
            self.b.add(temp)
            self.b_sum+=temp

    def median(self):
        self.adjust()
        return self.a[-1]
    def cost(self):
        med=self.median()
        return (med*len(self.a)-self.a_sum)+(self.b_sum-med*len(self.b))


obj=map()
for i in range(k):
    obj.insert(arr[i])
# median=obj.median()
print(obj.cost(),end=" ")
for i in range(k,n):
    obj.remove(arr[i-k])
    obj.insert(arr[i])
    print(obj.cost(),end=" ")

File: cses/sliding_window/sliding_window_cost.cpp
#include <bits/stdc++.h>
using namespace std;

class Map {
    multiset<int> low, high;
    long long sum_low = 0, sum_high = 0;

    void balance() {
        while (high.size() > low.size()) {
            int val = *high.begin();
            high.erase(high.begin());
            sum_high -= val;
            low.insert(val);
            sum_low += val;
        }
        while (low.size() > high.size() + 1) {
            auto it = prev(low.end());
            int val = *it;
            low.erase(it);
            sum_low -= val;
            high.insert(val);
            sum_high += val;
        }
    }

public:
    void insert(int val) {
        if (!low.empty() && val > *prev(low.end())) {
            high.insert(val);
            sum_high += val;
        } else {
            low.insert(val);
            sum_low += val;
        }
        balance();
    }

    void remove(int val) {
        if (!low.empty() && val <= *prev(low.end())) {
            auto it = low.find(val);
            if (it != low.end()) {
                low.erase(it);
                sum_low -= val;
            }
        } else {
            auto it = high.find(val);
            if (it != high.end()) {
                high.erase(it);
                sum_high -= val;
            }
        }
        balance();
    }

    int median() {
        balance();
        return *prev(low.end());
    }

    long long cost() {
        int med = median();
        return 1LL * med * (int)low.size() - sum_low + sum_high - 1LL * med * (int)high.size();
    }
};

int main() {
    int n, k;
    cin >> n >> k;
    vector<int> arr(n);
    for (int i = 0; i < n; ++i) cin >> arr[i];

    Map obj;
    for (int i = 0; i < k; ++i) obj.insert(arr[i]);
    cout << obj.cost() << " ";

    for (int i = k; i < n; ++i) {
        obj.remove(arr[i - k]);
        obj.insert(arr[i]);
        cout << obj.cost() << " ";
    }
    cout << endl;
    return 0;
}
